### DISCUSSION TITLE
Multicast Group Membership issue

### DISCUSSION THREAD

**Discussion by dereknbartram** [micropython/micropython#10905] [RP2040 / Pico] [answered]
I'm trying to write a micropython port of some code used to allow Amazon Alexa devices to control a RP2040 by pretending to be a Philips Hue bulb - the logic of how to do this is straightforward, however I'm having issues getting the raw UDP packets...

The piece causing the problem is the multicast address (239.255.255.250); packets sent to 255.255.255.255 are received, but not those to 239.255.255.250. This problem has been known for a while, and has been documented in https://github.com/micropython/micropython/issues/9105 and 'fixed' in https://github.com/georgerobotics/cyw43-driver/pull/25. The solution hasn't made it back into micropython core and appears to have stalled in development.

From a starting point of a clean and buildable micropython I have tried:

```
cd ports/rp2
make BOARD=PICO_W submodules
make BOARD=PICO_W clean
cd lib/cyw43-driver/
git fetch origin pull/25/head:multicast-register
git checkout multicast-register
cd ../..
cd ports/rp2
make BOARD=PICO_W
```

This results in the build error:

```
/home/derek_bartram/micropython/extmod/network_cyw43.c: In function 'network_cyw43_connect':
/home/derek_bartram/micropython/extmod/network_cyw43.c:229:67: error: 'CYW43_CHANNEL_NONE' undeclared (first use in this function)
  229 |         { MP_QSTR_channel, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = CYW43_CHANNEL_NONE} },
      |                                                                   ^~~~~~~~~~~~~~~~~~
/home/derek_bartram/micropython/extmod/network_cyw43.c:229:67: note: each undeclared identifier is reported only once for each function it appears in
make[3]: *** [CMakeFiles/firmware.dir/build.make:2441: CMakeFiles/firmware.dir/home/derek_bartram/micropython/extmod/network_cyw43.c.obj] Error 1
make[2]: *** [CMakeFiles/Makefile2:1498: CMakeFiles/firmware.dir/all] Error 2
make[1]: *** [Makefile:91: all] Error 2
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [Makefile:46: all] Error 1
```

Has anyone had any joy on building for PR25 please? And if so, how... or is there a copy someone has please?

The following is the basic replication of the issue: running it on a network with an Alexa and calling 'discover devices' should result in print output.
```py
ssid_ = 'SSID HERE'
wpa2_pass = 'PASSWORD HERE'

def do_connect():
    import network

    sta_if = network.WLAN(network.STA_IF)
    if not sta_if.isconnected():
        print("connecting to network...")
        sta_if.active(True)
        sta_if.connect(ssid_, wpa2_pass)
        while not sta_if.isconnected():
            pass
    print("network config:", sta_if.ifconfig())

do_connect()

import socket
try:
    from socket import sockaddr
except:
    sockaddr = lambda x: x

def inet_aton(addr):
    ip_as_bytes = bytes(map(int, addr.split(".")))
    return ip_as_bytes

UPNP_MCAST_IP = "239.255.255.250"
UPNP_PORT = 1900
BIND_IP = "0.0.0.0"
REUSE_SOCKET = 0

serv_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
addr = socket.getaddrinfo(BIND_IP, UPNP_PORT, socket.AF_INET, socket.SOCK_DGRAM)[0][4]
serv_sock.bind(addr)
serv_sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, inet_aton(UPNP_MCAST_IP) + inet_aton(BIND_IP));
if REUSE_SOCKET:
    resp_sock = serv_sock
else:
    resp_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

while True:
    print('Waiting for M-SEARCH...')
    data, addr = resp_sock.recvfrom(1024)
    if data.startswith(b"M-SEARCH"):
        print("From: ", sockaddr(addr))
        print(data)
```

**Comment by dpgeorge** [comment, marked as answer]
> Has anyone had any joy on building for PR25 please? And if so, how... or is there a copy someone has please?

That PR25 has already been merged to cyw43-driver (you can tell by the purple merged label at the top of that PR), and the submodule in the micropython repo updated to include that commit.

So, if you just build PICO_W at the current master commit of micropython, it will include that IGMP fix.

That said, there may still need to be some extra changes to make it work on the `rp2` port.

**Reply by dpgeorge** [reply]
I just tested multicast with micropython at 2bcd88d55645f2ea702ff70336b33be4661c97a8 (current master) and it works.

Here is a multicast sender:
```py
import socket

MULTICAST_IP = "239.255.255.250"
MULTICAST_PORT = 1900

s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
if hasattr(socket, "IP_MULTICAST_TTL"):
    s.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, 2)
for i in range(2):
    print(s.sendto(b'some data', (MULTICAST_IP, MULTICAST_PORT)))
s.close()
print('done')
```
Here is the corresponding receiver:
```py
import struct, socket

MULTICAST_IP = "239.255.255.250"
MULTICAST_PORT = 1900
MULTICAST_BIN = bytes(map(int, MULTICAST_IP.split("."))) + bytes(4)

s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
s.bind(("0.0.0.0", MULTICAST_PORT))

print('add membership')
s.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, MULTICAST_BIN)

for i in range(4):
    print(i, s.recv(1024))

print('drop membership')
s.setsockopt(socket.IPPROTO_IP, socket.IP_DROP_MEMBERSHIP, MULTICAST_BIN)
```
The above samples work on CPython, and MicroPython running on PICO_W and PYBD_SF2 (an STM32 board with the same WiFi chip and driver as the PICO_W).  Start the receiver first then run the sender.

**Reply by dereknbartram** [reply]
**Thank you dpgeorge** - completely user error on my part then. I've tested the above (receiver predominately) and yes, that 100% works. No issues on PI Pico W

Doing a little digging into why the original code didn't/no longer worked, but that's something I can sort.

**Comment by krisek** [comment]
> Here is the corresponding receiver:
> 
> ```python
> import struct, socket
> 
> MULTICAST_IP = "239.255.255.250"
> MULTICAST_PORT = 1900
> MULTICAST_BIN = bytes(map(int, MULTICAST_IP.split("."))) + bytes(4)
> 
> s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
> s.bind(("0.0.0.0", MULTICAST_PORT))
> 
> print('add membership')
> s.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, MULTICAST_BIN)
> 
> for i in range(4):
>     print(i, s.recv(1024))
> 
> print('drop membership')
> s.setsockopt(socket.IPPROTO_IP, socket.IP_DROP_MEMBERSHIP, MULTICAST_BIN)
> ```
> 

I tried to run the receiver on my Pico W and it doesn't receive anything. I tried with both latest stable and nightly MicroPython. And ideas what can be the issue? IP connectivity is OK, the board can be pinged.

**Reply by jimmo** [reply]
@krisek Please provide more information -- what are you trying to receive from? Does the same thing work if you run it in CPython on your PC?

**Comment by krisek** [comment]
> what are you trying to receive from

This is a super simple sender

```
import socket
import struct
import sys
import time

message = b'very important data'
multicast_group = ('239.255.255.250', 1901)
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

try:
    while True:
        print('sending {!r}'.format(message))
        sent = sock.sendto(message, multicast_group)

            
        time.sleep(5)
finally:
    print('closing socket')
    sock.close()
```

I see the messages coming in on systems with CPython on the same network, with this script:

```
import socket
import sys
import time

def main(argv):
    multicast_group = '239.255.255.250'
    multicast_port = 1901
    interface_ip = '0.0.0.0'

    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
    sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, 32)
    sock.bind(('', multicast_port))
    sock.setsockopt(socket.SOL_IP, socket.IP_ADD_MEMBERSHIP,
                    socket.inet_aton(multicast_group) + socket.inet_aton(interface_ip))

    while True:
        data, address = sock.recvfrom(1024)
        current_time = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())
        print(current_time)
        print(f"Received from {address}: {data.decode()}")


if __name__ == '__main__':
    main(sys.argv)
```

I cannot get it working on the Pico W with the above script (obviously I changed the port in the example script)... maybe I should move to an other multicast IP, to avoid any confusion with UPnP?

**Reply by krisek** [reply]
Actually that was it... by moving to 239.255.255.211, it all works now!

**Reply by krisek** [reply]
little wonder for this day

**Comment by IhorNehrutsa** [comment]
docs/esp32: UDP MULTICAST client/server example. #13166
